Modify deduplication cache in GlyphDisplayListCache to improve ~GlyphDisplayListCacheEntry performance

NathanSolomon733 opened the pull request:
#### 3c4ce68180d4becf33d07a339540892983c47c93
<pre>
Modify deduplication cache in GlyphDisplayListCache to improve ~GlyphDisplayListCacheEntry performance
<a href="https://rdar.apple.com/165895591">rdar://165895591</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=303610">https://bugs.webkit.org/show_bug.cgi?id=303610</a>

Reviewed by Kimmo Kinnunen.

The deduplication cache within the GlyphDisplayListCache used to share GlyphDisplayListCacheEntrys
has a large impact on SP3 performance. The cache has a hit rate of 26% in SP3, but it has a hit rate of 99% on MotionMark.
Removing the cache altogether significantly improves performance on SP3 by
reducing the entry removal work done in the destructor of GlyphDisplayListCacheEntry,
which is frequent in SP3.
In both benchmarks there is only a small working group
of reused GlyphDisplayListCacheEntrys. Taking advantage of this observation
and change the implementation of this cache to have a much smaller working set and remove the
need for manual removal of a GlyphDisplayListCacheEntry on GlyphDisplayListCacheEntry destruction.
This speeds up SP3 while maintaining the high cache hit rate seen in MotionMark.

* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/rendering/GlyphDisplayListCache.cpp:
(WebCore::GlyphDisplayListCacheKeyTranslator::equal):
(WebCore::GlyphDisplayListCache::getDisplayList):
(WebCore::GlyphDisplayListCacheEntry::~GlyphDisplayListCacheEntry): Deleted.
* Source/WebCore/rendering/GlyphDisplayListCache.h:
(WebCore::GlyphDisplayListCacheEntryHash::hash): Deleted.
(WebCore::GlyphDisplayListCacheEntryHash::equal): Deleted.

Canonical link: <a href="https://commits.webkit.org/304313@main">https://commits.webkit.org/304313@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ec5fb1ae57021d554ae19f150f9bc9cfe3b7cd15

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135143 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7535 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46401 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142649 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86937 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137012 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8170 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7384 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103265 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/70492 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138089 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5816 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121117 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84121 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5612 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3220 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3243 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114827 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39293 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145347 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7219 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39861 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111640 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7262 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6044 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112006 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28438 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5448 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117412 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/61191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7272 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35560 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7028 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70824 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7250 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7130 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebCore/rendering/GlyphDisplayListCache.cpp:
```diff
@@ -117,19 +119,28 @@ RefPtr<const DisplayList::DisplayList> GlyphDisplayListCache::getDisplayList(con
-        if (canShareDisplayList(result))
-            m_entries.add(entry.get());
+        if (canShareDisplayList(result)) {
+            m_entries.append(entry.get());
```
since 20 is such a low number, it might make sense to just reserveCapacity(s_max... + 1) here, so either we spend 0 memory or we spend max amount of entries.
Doesn't really make a difference, so might not be worth resubmitting. Just an observation.

NathanSolomon733 replied:
I can do this in a follow up PR. I will A/B test if reserving capacity shows to make some significant perf progression.

NathanSolomon733 committed:
Modify deduplication cache in GlyphDisplayListCache to improve ~GlyphDisplayListCacheEntry performance
rdar://165895591
https://bugs.webkit.org/show_bug.cgi?id=303610

Reviewed by Kimmo Kinnunen.

The deduplication cache within the GlyphDisplayListCache used to share GlyphDisplayListCacheEntrys
has a large impact on SP3 performance. The cache has a hit rate of 26% in SP3, but it has a hit rate of 99% on MotionMark.
Removing the cache altogether significantly improves performance on SP3 by
reducing the entry removal work done in the destructor of GlyphDisplayListCacheEntry,
which is frequent in SP3.
In both benchmarks there is only a small working group
of reused GlyphDisplayListCacheEntrys. Taking advantage of this observation
and change the implementation of this cache to have a much smaller working set and remove the
need for manual removal of a GlyphDisplayListCacheEntry on GlyphDisplayListCacheEntry destruction.
This speeds up SP3 while maintaining the high cache hit rate seen in MotionMark.

* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/rendering/GlyphDisplayListCache.cpp:
(WebCore::GlyphDisplayListCacheKeyTranslator::equal):
(WebCore::GlyphDisplayListCache::getDisplayList):
(WebCore::GlyphDisplayListCacheEntry::~GlyphDisplayListCacheEntry): Deleted.
* Source/WebCore/rendering/GlyphDisplayListCache.h:
(WebCore::GlyphDisplayListCacheEntryHash::hash): Deleted.
(WebCore::GlyphDisplayListCacheEntryHash::equal): Deleted.

Canonical link: https://commits.webkit.org/304313@main
